File: app/src/main/java/com/healthtracker/blood/suger/utils/FeedbackUtils.java
//package com.healthtracker.blood.suger.utils;
//
//import android.content.Context;
//import android.content.Intent;
//import android.content.pm.PackageManager;
//import android.net.Uri;
//import android.os.Build;
//import android.text.TextUtils;
//import android.widget.Toast;
//
//import com.healthtracker.blood.suger.BuildConfig;
//import com.healthtracker.blood.suger.R;
//
//import java.util.List;
//import java.util.Locale;
//import java.util.TimeZone;
//
//public class FeedbackUtils {
//
//    public static void sendFeedback(Context context, String feedback, List<Uri> fileList, String subAppend ){
//        StringBuilder sb = new StringBuilder();
//        String email = BuildConfig.FEEDBACK_EMAIL;
//        String app_version = "";
//        try {
//            app_version = context.getApplicationContext().getPackageManager().getPackageInfo(context.getPackageName(), 0).versionName + "(" + BuildConfig.VERSION_CODE + ")";
//        } catch (PackageManager.NameNotFoundException e) {
//            e.printStackTrace();
//        }
//        sb.append("Device info:");
//        sb.append("\n");
//        sb.append(app_version);
//        sb.append(", ").append(Build.MODEL);
//        sb.append(", OS_").append(Build.VERSION.RELEASE);
//        sb.append(", ");
//        sb.append(context.getResources().getDisplayMetrics().widthPixels).append("x").append(context.getResources().getDisplayMetrics().heightPixels);
//        sb.append(", ");
//        sb.append(context.getResources().getDisplayMetrics().densityDpi).append("Dpi");
//        sb.append(", ");
//        Locale locale = context.getResources().getConfiguration().locale;
//        sb.append(locale.getLanguage()).append(" _ ").append(locale.getCountry());
//        sb.append(", ");
//        sb.append(TimeZone.getDefault().getDisplayName(false, TimeZone.SHORT));
//        sb.append(", ");
//        sb.append("\n");
//        if (!TextUtils.isEmpty(title)) {
//            sb.append(title);
//            sb.append("\n");
//        }
//        if (!TextUtils.isEmpty(content)) {
//            sb.append(content);
//            sb.append("\n");
//        }
//
//        try {
//            Intent intent = new Intent(Intent.ACTION_SEND_MULTIPLE);
//            intent.setType("application/octet-stream");
//            intent.putExtra(Intent.EXTRA_EMAIL, new String[]{BuildConfig.FEEDBACK_EMAIL});
//            String subject = context.getString(R.string.feedback_email_title, context.getString(R.string.app_name));
////            if (!TextUtils.isEmpty(subAppend)) {
////                subject = subject + " " + subAppend;
////            }
//            intent.putExtra(Intent.EXTRA_SUBJECT, subject);
//            intent.putExtra(Intent.EXTRA_TEXT, sb.toString());
//            intent.putParcelableArrayListExtra(Intent.EXTRA_STREAM, );
//            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
//            if (EmailUtils.getInstance().hasGmail(context)) {
//                intent.setPackage(EmailUtils.PACKAGE_GMAIL);
//            } else if (EmailUtils.getInstance().hasEmailApp(context)) {
//                intent.setPackage(EmailUtils.PACKAGE_EMAIL_APP);
//            }
//
//            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
//                intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
//
//            } else {
//                intent.putExtra(Intent.EXTRA_TEXT, sb.toString());
//            }
//
//            context.startActivity(intent);
//        } catch (Throwable e) {
//            try {
//                e.printStackTrace();
//                Intent intent = new Intent(Intent.ACTION_SEND_MULTIPLE);
//                intent.setType("application/octet-stream");
//                intent.putExtra(Intent.EXTRA_EMAIL, new String[]{email});
//                String subject = context.getString(R.string.feedback_email_title, context.getString(R.string.app_name));
//                intent.putExtra(Intent.EXTRA_SUBJECT, subject);
//                intent.putExtra(Intent.EXTRA_TEXT, sb.toString());
//                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
//                intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
//                context.startActivity(Intent.createChooser(intent, subject));
//            } catch (Throwable e2) {
//                Toast.makeText(context.getApplicationContext(), context.getString(R.string.feedback_sending_failed), Toast.LENGTH_SHORT).show();
//            }
//        }
//    }
//
////    public static void uploadLogFile(Context context,int type) {
////        try {
//////            context = MusicFileProvider.Companion.getDPContext(context);
////            String src = LogUtils.INSTANCE.initLogFolder(context);
////            String dest = context.getCacheDir().getPath() + "/log.zip";
////            FileUtil.zip(src,dest);
////            File zipFile = new File(dest);
////            if (!zipFile.exists()) {
////                return;
////            }
////            Server.INSTANCE.uploadFile(zipFile,type);
////        }catch (Exception e){
////            e.printStackTrace();
////        }
////    }
//}

File: app/src/main/java/com/healthtracker/blood/suger/utils/FeedbackUtils.kt
package com.healthtracker.blood.suger.utils

import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.widget.Toast
import com.healthtracker.blood.suger.App
import com.healthtracker.blood.suger.BuildConfig
import com.healthtracker.blood.suger.R
import java.util.Locale
import java.util.TimeZone

/**
 * 反馈工具类
 * 用于发送用户反馈邮件，包含设备信息和附件
 */
object FeedbackUtils {

    /**
     * 发送反馈邮件
     * @param context 上下文
     * @param feedback 用户反馈内容
     * @param fileList 附件列表（图片 URI）
     * @param subAppend 邮件主题附加内容（可选）
     */
    fun sendFeedback(
        context: Context,
        feedback: String,
        fileList: List<Uri>,
        subAppend: String? = null,
        onResult:((Boolean) -> Unit)? = null

    ) {
        // 构建邮件正文
        val emailBody = buildEmailBody(context, feedback)
        
        // 获取反馈邮箱地址
        val email = "jiaoyun76861590@gmail.com"
        
        // 构建邮件主题
        val subject = buildSubject(context, subAppend)
        
        try {
            App.INSTANCE.isFeatureLeave = true
            // 第一次尝试：使用首选邮件应用
            sendEmailWithPreferredApp(context, email, subject, emailBody, fileList)
            onResult?.invoke(true)
        } catch (e: Throwable) {
            try {
                e.printStackTrace()
                // 第二次尝试：使用应用选择器
                sendEmailWithChooser(context, email, subject, emailBody, fileList)
                onResult?.invoke(true)
            } catch (e2: Throwable) {
                e2.printStackTrace()
                // 发送失败，显示提示
                onResult?.invoke(false)
            }
        }
    }

    /**
     * 构建邮件正文，包含设备信息和用户反馈
     */
    private fun buildEmailBody(context: Context, feedback: String): String {
        return buildString {
            // 设备信息部分
            append("Device info:\n")
            append(getAppVersion(context))
            append(", ${Build.MODEL}")
            append(", OS_${Build.VERSION.RELEASE}")
            
            // 屏幕信息
            val metrics = context.resources.displayMetrics
            append(", ${metrics.widthPixels}x${metrics.heightPixels}")
            append(", ${metrics.densityDpi}Dpi")
            
            // 语言和时区
            val locale = context.resources.configuration.locales[0]
            append(", ${locale.language}_${locale.country}")
            append(", ${TimeZone.getDefault().getDisplayName(false, TimeZone.SHORT)}")
            append("\n")
            
            // 用户反馈内容
            if (feedback.isNotBlank()) {
                append(feedback)
                append(":\n\n")
            }
        }
    }

    /**
     * 获取应用版本信息
     */
    private fun getAppVersion(context: Context): String {
        return try {
            val packageInfo = context.applicationContext.packageManager
                .getPackageInfo(context.packageName, 0)
            "${packageInfo.versionName}(${BuildConfig.VERSION_CODE})"
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
            "Unknown"
        }
    }

    /**
     * 构建邮件主题
     */
    private fun buildSubject(context: Context, subAppend: String?): String {
        val baseSubject = context.getString(
            R.string.feedback_email_title,
            context.getString(R.string.app_name)
        )
        return if (subAppend.isNullOrBlank()) {
            baseSubject
        } else {
            "$baseSubject $subAppend"
        }
    }

    /**
     * 使用首选邮件应用发送邮件
     */
    private fun sendEmailWithPreferredApp(
        context: Context,
        email: String,
        subject: String,
        body: String,
        fileList: List<Uri>
    ) {
        // 转换所有 file:// URI 为 content:// URI
        val convertedUris = fileList.map { convertFileUriToContentUri(context, it) }
        
        val intent = createEmailIntent(email, subject, body, convertedUris)
        
        // 设置首选邮件应用
        when {
            EmailUtils.getInstance().hasGmail(context) -> {
                intent.setPackage(EmailUtils.PACKAGE_GMAIL)
            }
            EmailUtils.getInstance().hasEmailApp(context) -> {
                intent.setPackage(EmailUtils.PACKAGE_EMAIL_APP)
            }
        }
        
        context.startActivity(intent)
    }

    /**
     * 使用应用选择器发送邮件
     */
    private fun sendEmailWithChooser(
        context: Context,
        email: String,
        subject: String,
        body: String,
        fileList: List<Uri>
    ) {
        // 转换所有 file:// URI 为 content:// URI
        val convertedUris = fileList.map { convertFileUriToContentUri(context, it) }
        
        val intent = createEmailIntent(email, subject, body, convertedUris)
        context.startActivity(Intent.createChooser(intent, subject))
    }

    /**
     * 将 file:// URI 转换为 content:// URI（通过 FileProvider）
     * 用于解决 Android N+ 的 FileUriExposedException 问题
     */
    private fun convertFileUriToContentUri(context: Context, uri: Uri): Uri {
        return if (uri.scheme == "file") {
            try {
                // file:// URI 需要通过 FileProvider 转换
                val file = java.io.File(uri.path ?: return uri)
                androidx.core.content.FileProvider.getUriForFile(
                    context,
                    "${context.packageName}.fileprovider",
                    file
                )
            } catch (e: Exception) {
                e.printStackTrace()
                uri // 转换失败，返回原 URI
            }
        } else {
            uri // content:// URI 直接返回
        }
    }

    /**
     * 创建邮件 Intent
     */
    private fun createEmailIntent(
        email: String,
        subject: String,
        body: String,
        fileList: List<Uri>
    ): Intent {
        return Intent(Intent.ACTION_SEND_MULTIPLE).apply {
            type = "application/octet-stream"
            putExtra(Intent.EXTRA_EMAIL, arrayOf(email))
            putExtra(Intent.EXTRA_SUBJECT, subject)
            putExtra(Intent.EXTRA_TEXT, body)
            
            // 添加附件
            if (fileList.isNotEmpty()) {
                putParcelableArrayListExtra(Intent.EXTRA_STREAM, ArrayList(fileList))
            }
            
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            
            // Android N+ 需要授予读取权限
            // 关键：FLAG_GRANT_READ_URI_PERMISSION 会自动应用到所有 EXTRA_STREAM 中的 URI
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            }
        }
    }
}
